x/ibc/02-client/keeper/client.go

```diff
@@ -101,45 +107,40 @@ func (k Keeper) UpdateClient(ctx sdk.Context, clientID string, header exported.H
 
 // CheckMisbehaviourAndUpdateState checks for client misbehaviour and freezes the
 // client if so.
-//
-// NOTE: In the first implementation, only Tendermint misbehaviour evidence is
-// supported.
-func (k Keeper) CheckMisbehaviourAndUpdateState(ctx sdk.Context, evidence evidenceexported.Evidence) error {
-	misbehaviour, ok := evidence.(tendermint.Misbehaviour)
-	if !ok {
-		return sdkerrors.Wrap(errors.ErrInvalidClientType, "consensus type is not Tendermint")
-	}
-
-	clientState, found := k.GetClientState(ctx, misbehaviour.ClientID)
+func (k Keeper) CheckMisbehaviourAndUpdateState(ctx sdk.Context, misbehaviour exported.Misbehaviour) error {
+	clientState, found := k.GetClientState(ctx, misbehaviour.GetClientID())
 	if !found {
-		return sdkerrors.Wrap(errors.ErrClientNotFound, misbehaviour.ClientID)
+		return sdkerrors.Wrap(types.ErrClientNotFound, misbehaviour.GetClientID())
 	}
 
-	committer, found := k.GetCommitter(ctx, misbehaviour.ClientID, uint64(misbehaviour.GetHeight()))
+	consensusState, found := k.GetConsensusState(ctx, misbehaviour.GetClientID(), uint64(misbehaviour.GetHeight()))
```

Comment:
The consensusState may not exist for the height where the misbehavior is found.

For example, if there were client state updates at height 10 and height 20, then there will be consensus states in the store at height 10 and 20

However, a misbehavior may have first happened at height 15 forwhich there is no consensus state stored. In this case, the `fromValidatorSet` should be the validatorSet from height 10.

And thus the `ConsensusState` passed in should also be from height 10

Reply:
ok, so change it to`min(clientState.GetLatestHeight(), misbehaviour.GetHeight())` = `min(20, 15)` but that won't work since it'll be still 15... there won't be a consensus state stored at that height. 

We'd need to iterate over the keys to get the latest height at which the consensus state was stored that satisfies `height <= min(clientState.GetLatestHeight(), misbehaviour.GetHeight())` and ` min(clientState.GetLatestHeight(), misbehaviour.GetHeight()) - height <= trusting period`

Reply:
Do you mean for the case where there was misbehaviour at height 15 but we are already at a height 20? If so, there should be an intermediary set of two conflicting headers, and if we have a consensus state before height 15 stored the evidence should be valid (since both will be valid updates to it). I don't think this requires changes (at least to the spec).

Reply:
> We'd need to iterate over the keys to get the latest height at which the consensus state 

Yes this is true. This is what the old `GetConsensusState` function did under the hood. You could reimplement this (either in `GetConsensusState` or in another helper function)

Or you could use the `iavlStore.ReverseIterator` and just grab the first element. This should be faster, but i couldn't find how to get this working on my first try hence issue #5365

Reply:
> Yup the evidence will still be valid because the update at 20 must have been valid. However, we still need to iterate through the store to find this past consensus state

Why do we need to iterate? We know what the old height was (from the `Evidence`).